File: algorithms.py
# Алгоритмы сортировки и поиска на Python

def selection_sort(arr):
    """Сортировка выбором - находит минимальный элемент и помещает его в начало"""
    n = len(arr)
    for i in range(n - 1):
        min_index = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_index]:
                min_index = j
        arr[i], arr[min_index] = arr[min_index], arr[i]
    return arr

def bubble_sort(arr):
    """Сортировка пузырьком - последовательно сравнивает и обменивает соседние элементы"""
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr

def insertion_sort(arr):
    """Сортировка вставками - вставляет каждый элемент в правильную позицию в отсортированной части"""
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

def merge_sort(arr):
    """Сортировка слиянием - рекурсивно делит массив и объединяет отсортированные части"""
    if len(arr) <= 1:
        return arr
    
    def merge(left, right):
        result = []
        i = j = 0
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1
        result.extend(left[i:])
        result.extend(right[j:])
        return result
    
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def shell_sort(arr):
    """Сортировка Шелла - улучшенная сортировка вставками с уменьшающимся шагом"""
    n = len(arr)
    gap = n // 2
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]
            j = i
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap
            arr[j] = temp
        gap //= 2
    return arr

def quick_sort(arr):
    """Быстрая сортировка - выбирает опорный элемент и рекурсивно сортирует части"""
    if len(arr) <= 1:
        return arr
    
    def partition(array, low, high):
        pivot = array[high]
        i = low - 1
        for j in range(low, high):
            if array[j] <= pivot:
                i += 1
                array[i], array[j] = array[j], array[i]
        array[i + 1], array[high] = array[high], array[i + 1]
        return i + 1
    
    def quick_sort_helper(array, low, high):
        if low < high:
            pi = partition(array, low, high)
            quick_sort_helper(array, low, pi - 1)
            quick_sort_helper(array, pi + 1, high)
    
    quick_sort_helper(arr, 0, len(arr) - 1)
    return arr

def heap_sort(arr):
    """Пирамидальная сортировка - строит кучу и извлекает максимальные элементы"""
    def heapify(array, n, i):
        largest = i
        left = 2 * i + 1
        right = 2 * i + 2
        
        if left < n and array[left] > array[largest]:
            largest = left
            
        if right < n and array[right] > array[largest]:
            largest = right
            
        if largest != i:
            array[i], array[largest] = array[largest], array[i]
            heapify(array, n, largest)
    
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)
    
    return arr

def linear_search(arr, target):
    """Линейный поиск - последовательно проверяет каждый элемент"""
    for i in range(len(arr)):
        if arr[i] == target:
            return i
    return -1

def binary_search(arr, target):
    """Бинарный поиск - работает только с отсортированными массивами"""
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

def interpolation_search(arr, target):
    """Интерполяционный поиск - предсказывает позицию элемента"""
    low, high = 0, len(arr) - 1
    
    while low <= high and target >= arr[low] and target <= arr[high]:
        if low == high:
            if arr[low] == target:
                return low
            return -1
        
        pos = low + ((high - low) // (arr[high] - arr[low])) * (target - arr[low])
        
        if arr[pos] == target:
            return pos
        elif arr[pos] < target:
            low = pos + 1
        else:
            high = pos - 1
    
    return -1

def fibonacci_search(arr, target):
    """Поиск Фибоначчи - использует числа Фибоначчи для определения позиций"""
    n = len(arr)
    fib_m2 = 0
    fib_m1 = 1
    fib_m = fib_m2 + fib_m1
    
    while fib_m < n:
        fib_m2 = fib_m1
        fib_m1 = fib_m
        fib_m = fib_m2 + fib_m1
    
    offset = -1
    
    while fib_m > 1:
        i = min(offset + fib_m2, n - 1)
        
        if arr[i] < target:
            fib_m = fib_m1
            fib_m1 = fib_m2
            fib_m2 = fib_m - fib_m1
            offset = i
        elif arr[i] > target:
            fib_m = fib_m2
            fib_m1 = fib_m1 - fib_m2
            fib_m2 = fib_m - fib_m1
        else:
            return i
    
    if fib_m1 and offset + 1 < n and arr[offset + 1] == target:
        return offset + 1
    
    return -1

# Демонстрация работы алгоритмов
if __name__ == "__main__":
    # Тестовые данные
    test_array = [64, 34, 25, 12, 22, 11, 90]
    sorted_array = [11, 12, 22, 25, 34, 64, 90]
    search_target = 25
    
    print("Исходный массив:", test_array)
    print("Отсортированный массив:", sorted_array)
    print("Цель поиска:", search_target)
    print()
    
    # Демонстрация сортировок
    print("=== АЛГОРИТМЫ СОРТИРОВКИ ===")
    
    arr1 = test_array.copy()
    print(f"Сортировка выбором: {selection_sort(arr1)}")
    
    arr2 = test_array.copy()
    print(f"Сортировка пузырьком: {bubble_sort(arr2)}")
    
    arr3 = test_array.copy()
    print(f"Сортировка вставками: {insertion_sort(arr3)}")
    
    arr4 = test_array.copy()
    print(f"Сортировка слиянием: {merge_sort(arr4)}")
    
    arr5 = test_array.copy()
    print(f"Сортировка Шелла: {shell_sort(arr5)}")
    
    arr6 = test_array.copy()
    print(f"Быстрая сортировка: {quick_sort(arr6)}")
    
    arr7 = test_array.copy()
    print(f"Пирамидальная сортировка: {heap_sort(arr7)}")
    
    print("\n=== АЛГОРИТМЫ ПОИСКА ===")
    
    print(f"Линейный поиск {search_target}: {linear_search(sorted_array, search_target)}")
    print(f"Бинарный поиск {search_target}: {binary_search(sorted_array, search_target)}")
    print(f"Интерполяционный поиск {search_target}: {interpolation_search(sorted_array, search_target)}")
    print(f"Поиск Фибоначчи {search_target}: {fibonacci_search(sorted_array, search_target)}")

File: algorithms.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Сортировка выбором - находит минимальный элемент и помещает его в начало
void selectionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        int minIndex = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        swap(arr[i], arr[minIndex]);
    }
}

// Сортировка пузырьком - последовательно сравнивает и обменивает соседние элементы
void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

// Сортировка вставками - вставляет каждый элемент в правильную позицию в отсортированной части
void insertionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// Вспомогательная функция для сортировки слиянием
void merge(vector<int>& arr, int left, int mid, int right) {
    vector<int> temp(right - left + 1);
    int i = left, j = mid + 1, k = 0;
    
    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }
    
    while (i <= mid) temp[k++] = arr[i++];
    while (j <= right) temp[k++] = arr[j++];
    
    for (int p = 0; p < k; p++) {
        arr[left + p] = temp[p];
    }
}

// Сортировка слиянием - рекурсивно делит массив и объединяет отсортированные части
void mergeSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// Сортировка Шелла - улучшенная сортировка вставками с уменьшающимся шагом
void shellSort(vector<int>& arr) {
    int n = arr.size();
    for (int gap = n / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < n; i++) {
            int temp = arr[i];
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                arr[j] = arr[j - gap];
            }
            arr[j] = temp;
        }
    }
}

// Вспомогательная функция для быстрой сортировки
int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    
    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Быстрая сортировка - выбирает опорный элемент и рекурсивно сортирует части
void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Вспомогательная функция для пирамидальной сортировки
void heapify(vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;
    
    if (left < n && arr[left] > arr[largest])
        largest = left;
    
    if (right < n && arr[right] > arr[largest])
        largest = right;
    
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

// Пирамидальная сортировка - строит кучу и извлекает максимальные элементы
void heapSort(vector<int>& arr) {
    int n = arr.size();
    
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);
    
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

// Линейный поиск - последовательно проверяет каждый элемент
int linearSearch(const vector<int>& arr, int target) {
    for (int i = 0; i < arr.size(); i++) {
        if (arr[i] == target) {
            return i;
        }
    }
    return -1;
}

// Бинарный поиск - работает только с отсортированными массивами
int binarySearch(const vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;
    
    while (left <= right) {
        int mid = left + (right - left) / 2;
        
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

// Интерполяционный поиск - предсказывает позицию элемента
int interpolationSearch(const vector<int>& arr, int target) {
    int low = 0;
    int high = arr.size() - 1;
    
    while (low <= high && target >= arr[low] && target <= arr[high]) {
        if (low == high) {
            if (arr[low] == target) return low;
            return -1;
        }
        
        int pos = low + (((double)(high - low) / 
                  (arr[high] - arr[low])) * (target - arr[low]));
        
        if (arr[pos] == target)
            return pos;
        else if (arr[pos] < target)
            low = pos + 1;
        else
            high = pos - 1;
    }
    return -1;
}

// Поиск Фибоначчи - использует числа Фибоначчи для определения позиций
int fibonacciSearch(const vector<int>& arr, int target) {
    int n = arr.size();
    
    int fibM2 = 0; // F(m-2)
    int fibM1 = 1; // F(m-1)
    int fibM = fibM2 + fibM1; // F(m)
    
    while (fibM < n) {
        fibM2 = fibM1;
        fibM1 = fibM;
        fibM = fibM2 + fibM1;
    }
    
    int offset = -1;
    
    while (fibM > 1) {
        int i = min(offset + fibM2, n - 1);
        
        if (arr[i] < target) {
            fibM = fibM1;
            fibM1 = fibM2;
            fibM2 = fibM - fibM1;
            offset = i;
        } else if (arr[i] > target) {
            fibM = fibM2;
            fibM1 = fibM1 - fibM2;
            fibM2 = fibM - fibM1;
        } else {
            return i;
        }
    }
    
    if (fibM1 && offset + 1 < n && arr[offset + 1] == target)
        return offset + 1;
    
    return -1;
}

// Функция для вывода массива
void printArray(const vector<int>& arr) {
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main() {
    // Тестовые данные
    vector<int> testArray = {64, 34, 25, 12, 22, 11, 90};
    vector<int> sortedArray = {11, 12, 22, 25, 34, 64, 90};
    int searchTarget = 25;
    
    cout << "Исходный массив: ";
    printArray(testArray);
    cout << "Отсортированный массив: ";
    printArray(sortedArray);
    cout << "Цель поиска: " << searchTarget << endl << endl;
    
    // Демонстрация сортировок
    cout << "=== АЛГОРИТМЫ СОРТИРОВКИ ===" << endl;
    
    vector<int> arr1 = testArray;
    selectionSort(arr1);
    cout << "Сортировка выбором: ";
    printArray(arr1);
    
    vector<int> arr2 = testArray;
    bubbleSort(arr2);
    cout << "Сортировка пузырьком: ";
    printArray(arr2);
    
    vector<int> arr3 = testArray;
    insertionSort(arr3);
    cout << "Сортировка вставками: ";
    printArray(arr3);
    
    vector<int> arr4 = testArray;
    mergeSort(arr4, 0, arr4.size() - 1);
    cout << "Сортировка слиянием: ";
    printArray(arr4);
    
    vector<int> arr5 = testArray;
    shellSort(arr5);
    cout << "Сортировка Шелла: ";
    printArray(arr5);
    
    vector<int> arr6 = testArray;
    quickSort(arr6, 0, arr6.size() - 1);
    cout << "Быстрая сортировка: ";
    printArray(arr6);
    
    vector<int> arr7 = testArray;
    heapSort(arr7);
    cout << "Пирамидальная сортировка: ";
    printArray(arr7);
    
    cout << "\n=== АЛГОРИТМЫ ПОИСКА ===" << endl;
    
    cout << "Линейный поиск " << searchTarget << ": " 
         << linearSearch(sortedArray, searchTarget) << endl;
    cout << "Бинарный поиск " << searchTarget << ": " 
         << binarySearch(sortedArray, searchTarget) << endl;
    cout << "Интерполяционный поиск " << searchTarget << ": " 
         << interpolationSearch(sortedArray, searchTarget) << endl;
    cout << "Поиск Фибоначчи " << searchTarget << ": " 
         << fibonacciSearch(sortedArray, searchTarget) << endl;
    
    return 0;
}
